src/views/Simulation/components/PixiComponents/components/pixiNode.js

```diff
@@ -61,27 +64,203 @@ export function ensureStockNode(rootContainer, id, app, _textureCache) {
     const label = node.getChildByName('label');
 
     const v = Number(value ?? 0);
-    const r = v > 0 ? Math.min(STYLE.maxRadius, STYLE.minRadius + Math.log10(v + 1) * 10) : STYLE.coreIdleRadius;
 
-    drawCore(core, r, STYLE.coreFill, STYLE.ringColor, STYLE.ringWidth);
+    const baseRadius = v > 0 ? STYLE.minRadius + Math.log10(v + 1) * 10 : STYLE.minRadius;
+    const r = Math.min(STYLE.maxRadius, baseRadius);
 
+    const fillColor = v > 10 ? 0xff5555 : STYLE.coreFill;
     label.text = v > 0 ? String(v) : '';
-    label.style.fontSize = Math.max(10, r * 0.8);
-    label.position.set(0, 0);
+
+    drawCore(core, r, fillColor, label.text);
+
+    label.style.fontSize = Math.max(STYLE.fontSize, r * 1);
+    label.resolution = app.renderer.resolution || window.devicePixelRatio || 2;
+    label.scale.set(1 / 3);
 
     node.position.set(x, y);
-    node.scale.set(worldScale);
+
+    let scale = 1 / Math.pow(worldScale, 0.9);
+    scale = Math.max(0.25, Math.min(4.0, scale));
+    node.scale.set(scale);
   }
 
   return { container: node, update };
 }
 
-export function removeStockNode(rootContainer, id) {
-  const name = `node-${id}`;
-  const n = rootContainer.getChildByName(name);
-  if (n && n.parent) n.parent.removeChild(n);
+export function updateClusters({ app, mapContainer, prepared, layersRef }) {
+  const { links: linksLayer, nodes: nodesLayer } = layersRef.current;
+  if (!app || !nodesLayer || !prepared) return;
+
+  const worldScale = mapContainer?.worldTransform?.a ?? 1;
+  const pixelRatio = app.renderer?.resolution ?? window.devicePixelRatio ?? 1;
+  const liveNodeNames = new Set();
+
+  const haversineDistance = (lat1, lon1, lat2, lon2) => {
+    const R = 6371; // km
+    const dLat = ((lat2 - lat1) * Math.PI) / 180;
+    const dLon = ((lon2 - lon1) * Math.PI) / 180;
+    const a =
+      Math.sin(dLat / 2) ** 2 +
+      Math.cos((lat1 * Math.PI) / 180) * Math.cos((lat2 * Math.PI) / 180) * Math.sin(dLon / 2) ** 2;
+    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
+    return R * c;
+  };
+
+  const getLatLon = (n) => ({
+    lat: Number(n?.data?.latitude ?? n?.data?.lat),
```

Comment:
We guess `lat` and `lon` attributes significations because we know the context but it is better to avoid abreviations.
Maybe directly call this attributes `latitude` and `longitude`.